components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractLocalApplicationAuthenticator.java

```diff
@@ -0,0 +1,315 @@
+            } else {
+                AuthenticationDataPublisher authnDataPublisherProxy = FrameworkServiceDataHolder.getInstance()
+                        .getAuthnDataPublisherProxy();
+                Map<String, Object> unmodifiableParamMap;
+                String eventName;
+                try {
+                    if (eventFiringEnabledForAccountLocking()) {
+                        eventName = IdentityEventConstants.Event.PRE_AUTHENTICATION;
+                        fireEvent(context, eventName, false);
+                    }
+                    processAuthenticationResponse(request, response, context);
+                    if (this instanceof LocalApplicationAuthenticator && !context.getSequenceConfig()
+                            .getApplicationConfig().isSaaSApp()) {
+                        processWithNonSaaSApp(context);
+                    }
+                    request.setAttribute(FrameworkConstants.REQ_ATTR_HANDLED, true);
+                    context.setProperty(FrameworkConstants.LAST_FAILED_AUTHENTICATOR, null);
+                    if (authnDataPublisherProxy != null && authnDataPublisherProxy.isEnabled(context)) {
+                        unmodifiableParamMap = getStringObjectMap(context, context.getSubject());
+                        authnDataPublisherProxy.publishAuthenticationStepSuccess(request, context, unmodifiableParamMap);
+                        if (eventFiringEnabledForAccountLocking()) {
+                            eventName = IdentityEventConstants.Event.POST_AUTHENTICATION;
+                            fireEvent(context, eventName, true);
+                        }
+                    }
+                    return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+                } catch (AuthenticationFailedException e) {
+                    if (checkUserAccountStatus(response, context)) {
+                        return AuthenticatorFlowStatus.INCOMPLETE;
+                    }
+                    if (authnDataPublisherProxy != null && authnDataPublisherProxy.isEnabled(context)) {
+                        unmodifiableParamMap = getStringObjectMap(context, e.getUser());
+                        authnDataPublisherProxy.publishAuthenticationStepFailure(request, context, unmodifiableParamMap);
+                        if (eventFiringEnabledForAccountLocking()) {
+                            eventName = IdentityEventConstants.Event.POST_AUTHENTICATION;
+                            fireEvent(context, eventName, false);
+                        }
+                    }
+                    request.setAttribute(FrameworkConstants.REQ_ATTR_HANDLED, true);
+                    // Decide whether we need to redirect to the login page to retry authentication.
+                    boolean sendToMultiOptionPage =
+                            isStepHasMultiOption(context) && isRedirectToMultiOptionPageOnFailure();
+                    if (retryAuthenticationEnabled(context) && !sendToMultiOptionPage) {
```

Comment:
It seems we can still reduce much of these duplicates by introducing new protected methods at AbstractApplicationAuthenticator and organizing them. Can you please work on that? We should make the code duplication as minimal as possible.